File: Engine/ceCore/entity/camerastate.cc


#include <ceCore/entity/camerastate.hh>
#include <ceCore/math/matrix4f.hh>
#include <ceCore/math/vector3f.hh>
#include <ceCore/math/vector4f.hh>
#include <math.h>
#include <ceCore/graphics/irendertarget2d.hh>
#include <ceCore/graphics/postprocessing.hh>
#include <ceCore/graphics/scene/gfxcamera.hh>
#include <ceCore/graphics/scene/igfxscene.hh>
#include <ceCore/entity/world.hh>

namespace ce
{


CameraState::CameraState()
    : SpatialState(),
      m_near(1.0f),
      m_far(1024.0f),
      m_angle((float) M_PI / 4.0f),
      m_angleWidthHeight(0.0f),
      m_gfxCamera(new GfxCamera()),
      m_order(0),
      m_renderTarget(nullptr),
      m_postProcessing(nullptr)
{
  CE_CLASS_GEN_CONSTR;
  m_gfxCamera->SetCamera(&m_camera);
  m_gfxCamera->SetProjector(&m_projector);
  UpdateGfxCamera();
}

CameraState::~CameraState()
{
  m_gfxCamera->Release();
}

void CameraState::SetNear(float near)
{
  m_near = near;
  UpdateGfxCamera();
}

float CameraState::GetNear() const
{
  return m_near;
}

void CameraState::SetFar(float far)
{
  m_far = far;
  UpdateGfxCamera();
}

float CameraState::GetFar() const
{
  return m_far;
}

void CameraState::SetAngle(float angle)
{
  m_angle = angle;
  UpdateGfxCamera();
}

float CameraState::GetAngle() const
{
  return m_angle;
}

void CameraState::SetAngleWidthHeight(float angleWidthHeight)
{
  m_angleWidthHeight = angleWidthHeight;
  UpdateGfxCamera();
}

float CameraState::GetAngleWidthHeight() const
{
  return m_angleWidthHeight;
}

void CameraState::SetOrder(int order)
{
  m_order = order;
  UpdateGfxCamera();
}

int CameraState::GetOrder() const
{
  return m_order;
}


void CameraState::SetRenderTarget(iRenderTarget2D *renderTarget)
{
  CE_SET(m_renderTarget, renderTarget);
  UpdateGfxCamera();
}

iRenderTarget2D* CameraState::GetRenderTarget()
{
  return m_renderTarget;
}

const iRenderTarget2D* CameraState::GetRenderTarget() const
{
  return m_renderTarget;
}



void CameraState::SetClearMode(eClearMode clearMode)
{
  m_clearMode = clearMode;
  UpdateGfxCamera();
}

eClearMode CameraState::GetClearMode() const
{
  return m_clearMode;
}

void CameraState::SetClearColor(const Color4f &clearColor)
{
  m_clearColor = clearColor;
  UpdateGfxCamera();
}

const Color4f &CameraState::GetClearColor() const
{
  return m_clearColor;
}

void CameraState::SetClearDepth(float clearDepth)
{
  m_clearDepth = clearDepth;
  UpdateGfxCamera();
}

float CameraState::GetClearDepth() const
{
  return m_clearDepth;
}

void CameraState::SetRenderShadows(bool renderShadows)
{
  m_renderShadows = renderShadows;
  UpdateGfxCamera();
}

bool CameraState::IsRenderShadows() const
{
  return m_renderShadows;
}

void CameraState::SetPostProcessing(ce::PostProcessing *postProcessing)
{
  CE_SET(m_postProcessing, postProcessing);
  UpdateGfxCamera();
}

ce::PostProcessing* CameraState::GetPostPRocessing() const
{
  return m_postProcessing;
}

const Camera &CameraState::GetCamera() const
{
  return m_camera;
}

const Projector &CameraState::GetProjector() const
{
  return m_projector;
}


void CameraState::OnAttachedToWorld(World *world)
{
  iGfxScene *scene = world->GetScene();
  scene->Add(m_gfxCamera);
}

void CameraState::OnDetachedFromWorld(World *world)
{
  iGfxScene *scene = world->GetScene();
  scene->Remove(m_gfxCamera);
}

void CameraState::TransformationUpdatedPreChildren()
{
  const Matrix4f &m = GetGlobalMatrix();

  Vector3f eye(m * Vector4f(0.0, 0.0, 0.0, 1.0f));
  Vector3f dir(m * Vector4f(0.0, 0.0, 1.0, 0.0f));
  Vector3f up(m * Vector4f(0.0, 1.0, 0.0, 0.0f));

  m_camera.SetEye(eye);
  m_camera.SetSpot(eye + dir);
  m_camera.SetUp(up);
}

void CameraState::UpdateGfxCamera()
{
  m_gfxCamera->SetOrder(m_order);
  m_gfxCamera->SetRenderTarget(m_renderTarget);
  m_gfxCamera->SetRenderShadows(m_renderShadows);
  m_gfxCamera->SetPostProcessing(m_postProcessing);
  m_gfxCamera->UpdateData(m_near, m_far, m_angle, m_angleWidthHeight);
  m_gfxCamera->UpdateClear(m_clearMode, m_clearColor, m_clearDepth);
}

}

File: JCryoStorm/CsCore/src/main/java/org/cryo/core/entity/CameraState.java
package org.cryo.core.entity;

import org.cryo.core.CsClass;
import org.cryo.core.graphics.EClearColorMode;
import org.cryo.core.graphics.EClearMode;
import org.cryo.core.math.Color4f;

import static org.cryo.core.entity.CameraStateNative.*;

@CsClass(CS_CLASS_NAME)
public class CameraState extends EntityState {

    public CameraState() {
    }

    public CameraState(long ref) {
        super(ref);
    }


    public void setNear (float near) {
        nSetNear(getRef(), near);
    }

    public float getNear () {
        return nGetNear(getRef());
    }

    public void setFar (float far) {
        nSetFar(getRef(), far);
    }

    public float getFar () {
        return nGetFar(getRef());
    }

    public void setAngle (float angle) {
        nSetAngle(getRef(), angle);
    }

    public float getAngle () {
        return nGetAngle(getRef());
    }

    public void setAngleWidthHeight(float angleWidthHeight) {
        nSetAngleWidthHeight(getRef(), angleWidthHeight);
    }

    public float getAngleWidthHeight () {
        return nGetAngleWidthHeight(getRef());
    }

    public void setOrder (int order) {
        nSetOrder(getRef(), order);
    }

    public int getOrder () {
        return nGetOrder(getRef());
    }

    public void setClearMode (EClearMode clearMode) {
        nSetClearMode(getRef(), clearMode.ordinal());
    }

    public EClearMode getClearMode () {
        return EClearMode.values()[nGetClearMode(getRef())];
    }

    public void setClearColorMode(EClearColorMode clearColorMode) {
        nSetClearColorMode(getRef(), clearColorMode.ordinal());
    }

    public EClearColorMode getClearColorMode () {
        return EClearColorMode.values()[nGetClearColorMode(getRef())];
    }

    public void setClearColor (Color4f clearColor) {
        nSetClearColor(getRef(), clearColor.r, clearColor.g, clearColor.b, clearColor.a);
    }

    private final float[] clearColorMem = new float[4];
    private final Color4f clearColor = new Color4f();
    public Color4f getClearColor () {
        nGetClearColor(getRef(), clearColorMem);
        return clearColor.set(clearColorMem);
    }

    public void setClearDepth (float clearDepth) {
        nSetClearDepth(getRef(), clearDepth);
    }

    public float getClearDepth () {
        return nGetClearDepth(getRef());
    }

    public void setRenderShadows (boolean renderShadows) {
        nSetRenderShadows(getRef(), renderShadows);
    }

    public boolean isRenderShadows () {
        return nIsRenderShadows(getRef());
    }

    public void update (long width, long height) {
        nUpdate(getRef(), width, height);
    }
}
